# Prueba 2

**Nombre:** Julio Valdés
<br/>
**Generación:** GC

### Configuración y lectura de datos

In [1]:
from pyspark.sql import SparkSession

spark = SparkSession\
    .builder\
    .appName("prueba-2")\
    .enableHiveSupport()\
    .getOrCreate()

Starting Spark application


SparkSession available as 'spark'.


In [2]:
spark

In [66]:
business = spark.read.json('s3://bigdata-desafio/yelp-data/business.json')
business.show(5)

+--------------------+--------------------+--------------------+--------------------+-----------+--------------------+-------+-------------+---------------+--------------------+-----------+------------+-----+-----+
|             address|          attributes|         business_id|          categories|       city|               hours|is_open|     latitude|      longitude|                name|postal_code|review_count|stars|state|
+--------------------+--------------------+--------------------+--------------------+-----------+--------------------+-------+-------------+---------------+--------------------+-----------+------------+-----+-----+
|2818 E Camino Ace...|[,,,,,,,,,,,,,,,,...|1SWheh84yJXfytovI...|   Golf, Active Life|    Phoenix|                null|      0|   33.5221425|   -112.0184807|Arizona Biltmore ...|      85016|           5|  3.0|   AZ|
|30 Eglinton Avenue W|[,, u'full_bar', ...|QXAEGFB4oINsVuTFx...|Specialty Food, R...|Mississauga|[9:0-1:0, 9:0-0:0...|      1|43.6054989743|

In [4]:
review = spark.read.json('s3://bigdata-desafio/yelp-data/review.json')
review.show(5)

+--------------------+----+-------------------+-----+--------------------+-----+--------------------+------+--------------------+
|         business_id|cool|               date|funny|           review_id|stars|                text|useful|             user_id|
+--------------------+----+-------------------+-----+--------------------+-----+--------------------+------+--------------------+
|ujmEBvifdJM6h6RLv...|   0|2013-05-07 04:34:36|    1|Q1sbwvVQXV2734tPg...|  1.0|Total bill for th...|     6|hG7b0MtEbXx5QzbzE...|
|NZnhc2sEQy3RmzKTZ...|   0|2017-01-14 21:30:33|    0|GJXCdrto3ASJOqKeV...|  5.0|I *adore* Travis ...|     0|yXQM5uF2jS6es16SJ...|
|WTqjgwHlXbSFevF32...|   0|2016-11-09 20:09:03|    0|2TzJjDVDEuAW6MR5V...|  5.0|I have to say tha...|     3|n6-Gk65cPZL6Uz8qR...|
|ikCg8xy5JIg_NGPx-...|   0|2018-01-09 20:56:38|    0|yi0R0Ugj_xUx_Nek0...|  5.0|Went in for a lun...|     0|dacAIZ6fTM6mqwW5u...|
|b1b1eb3uo-w561D0Z...|   0|2018-01-30 23:07:38|    0|11a8sVPMUFtaC7_AB...|  1.0|Today was 

In [5]:
users = spark.read.json('s3://bigdata-desafio/yelp-data/user.json')
users.show(5)

+-------------+---------------+---------------+----------------+--------------+---------------+---------------+---------------+-----------------+----------------+------------------+-----------------+----+-------------------+----+--------------------+-----+------+------------+------+--------------------+-------------------+
|average_stars|compliment_cool|compliment_cute|compliment_funny|compliment_hot|compliment_list|compliment_more|compliment_note|compliment_photos|compliment_plain|compliment_profile|compliment_writer|cool|              elite|fans|             friends|funny|  name|review_count|useful|             user_id|      yelping_since|
+-------------+---------------+---------------+----------------+--------------+---------------+---------------+---------------+-----------------+----------------+------------------+-----------------+----+-------------------+----+--------------------+-----+------+------------+------+--------------------+-------------------+
|         4.03|          

In [6]:
checkin = spark.read.json('s3://bigdata-desafio/yelp-data/checkin.json')
checkin.show(5)

+--------------------+--------------------+
|         business_id|                date|
+--------------------+--------------------+
|--1UhMGODdWsrMast...|2016-04-26 19:49:...|
|--6MefnULPED_I942...|2011-06-04 18:22:...|
|--7zmmkVg-IMGaXbu...|2014-12-29 19:25:...|
|--8LPVSo5i0Oo61X0...| 2016-07-08 16:43:30|
|--9QQLMTbFzLJ_oT-...|2010-06-26 17:39:...|
+--------------------+--------------------+
only showing top 5 rows

In [7]:
tip = spark.read.json('s3://bigdata-desafio/yelp-data/tip.json')
tip.show(5)

+--------------------+----------------+-------------------+--------------------+--------------------+
|         business_id|compliment_count|               date|                text|             user_id|
+--------------------+----------------+-------------------+--------------------+--------------------+
|VaKXUpmWTTWDKbpJ3...|               0|2014-03-27 03:51:24|Great for watchin...|UPw5DWs_b-e2JRBS-...|
|OPiPeoJiv92rENwbq...|               0|2013-05-25 06:00:56|Happy Hour 2-4 da...|Ocha4kZBHb4JK0lOW...|
|5KheTjYPu1HcQzQFt...|               0|2011-12-26 01:46:17|Good chips and sa...|jRyO2V1pA4CdVVqCI...|
|TkoyGi8J7YFjA6Sba...|               0|2014-03-23 21:32:49|The setting and d...|FuTJWFYm4UKqewaos...|
|AkL6Ous6A1atZejfZ...|               0|2012-10-06 00:19:27|Molly is definate...|LUlKtaM3nXd-E4N4u...|
+--------------------+----------------+-------------------+--------------------+--------------------+
only showing top 5 rows

In [8]:
photo = spark.read.json('s3://bigdata-desafio/yelp-data/photo.json')
photo.show(5)

+--------------------+-------+------+--------------------+
|         business_id|caption| label|            photo_id|
+--------------------+-------+------+--------------------+
|rcaPajgKOJC2vo_l3...|       |inside|MllA1nNpcp1kDteVg...|
|Kn23LDd740SBVJ7mu...|       |inside|YjxBE88Bf6CmTEF2L...|
|ZkGDCVKSdf8m76cnn...|       |  food|1f7izSjM0WjkDRIVb...|
|bF8gv7k_rwZtiDLP2...|       |inside|NcSlcDTEEeOaixotO...|
|50Anorn0DJXFhBr9a...|       |inside|5IiIo5UKEW0lWqZ6s...|
+--------------------+-------+------+--------------------+
only showing top 5 rows

### Ejercicio 1: Identificando usuarios molestosos (12 puntos)

#### Identifique en una variable dummy todos los usuarios que se puedan clasificar como molestosos acorde al criterio

In [9]:
from pyspark.sql.functions import when

users = users\
            .withColumn('flag_molestoso',
                when((users.average_stars <= 2) | (users.review_count < 100) | (users.fans == 0), 1)\
                .otherwise(0))
users\
    .select('user_id','average_stars','review_count','fans', 'flag_molestoso')\
    .show(10)

+--------------------+-------------+------------+----+--------------+
|             user_id|average_stars|review_count|fans|flag_molestoso|
+--------------------+-------------+------------+----+--------------+
|l6BmjZMeQD3rDxWUb...|         4.03|          95|   5|             1|
|4XChL029mKr5hydo7...|         3.63|          33|   4|             1|
|bc8C_eETBWL0olvFS...|         3.71|          16|   0|             1|
|dD0gZpBctWGdWo9Wl...|         4.85|          17|   5|             1|
|MM4RJAeH6yuaN8oZD...|         4.08|         361|  39|             0|
|0rK89TS8xqy1wI4nY...|          4.2|         214| 186|             0|
|TEtzbpgA2BFBrC0y0...|         4.39|        1122| 696|             0|
|KGuqerdeNhxzXZEyB...|         4.33|           6|   0|             1|
|T0gWkTHWRChVUe_Dn...|         4.21|         859|  57|             0|
|NQffx45eJaeqhFcMa...|         4.53|         124|  15|             0|
+--------------------+-------------+------------+----+--------------+
only showing top 10 

#### Recodificaciones en el archivo user.json
* friends , que corresponde a un string con todos los user_id de otros usuarios quesiguen al usuario . El objetivo es contar la cantidad de amigos existentes.
* elite , que corresponde a un string con todos los años en los que el usuario fue considerado como un reviewer de elite. El objetivo es contar la cantidad de años en los cuales se consideró como elite.
* Asegúrese de eliminar los siguientes registros: friends , yelping_since , name , elite , user_id .

In [10]:
from pyspark.sql.functions import udf
from pyspark.sql.types import IntegerType

array_count = udf(lambda x: x.count(','), IntegerType())

users = users\
            .withColumn('friends_count', array_count(users.friends))\
            .withColumn('elite_count', array_count(users.elite))

users = users\
            .drop('friends', 'yelping_since', 'name', 'elite', 'user_id')
users.printSchema()

root
 |-- average_stars: double (nullable = true)
 |-- compliment_cool: long (nullable = true)
 |-- compliment_cute: long (nullable = true)
 |-- compliment_funny: long (nullable = true)
 |-- compliment_hot: long (nullable = true)
 |-- compliment_list: long (nullable = true)
 |-- compliment_more: long (nullable = true)
 |-- compliment_note: long (nullable = true)
 |-- compliment_photos: long (nullable = true)
 |-- compliment_plain: long (nullable = true)
 |-- compliment_profile: long (nullable = true)
 |-- compliment_writer: long (nullable = true)
 |-- cool: long (nullable = true)
 |-- fans: long (nullable = true)
 |-- funny: long (nullable = true)
 |-- review_count: long (nullable = true)
 |-- useful: long (nullable = true)
 |-- flag_molestoso: integer (nullable = false)
 |-- friends_count: integer (nullable = true)
 |-- elite_count: integer (nullable = true)

In [11]:
users.take(1)

[Row(average_stars=4.03, compliment_cool=1, compliment_cute=0, compliment_funny=1, compliment_hot=2, compliment_list=0, compliment_more=0, compliment_note=1, compliment_photos=0, compliment_plain=1, compliment_profile=0, compliment_writer=2, cool=25, fans=5, funny=17, review_count=95, useful=84, flag_molestoso=1, friends_count=98, elite_count=2)]

### Requerimientos

#### Genere la medición de usuarios molestos en base a los criterios expuestos. (2 puntos)

In [12]:
print("Cantidad de usuarios molestosos")
users\
    .where('flag_molestoso = 1')\
    .count()

Cantidad de usuarios molestosos
1566689

#### Divida la muestra en conjuntos de entrenamiento (preservando un 70% de los datos) y validación (preservando un 30% de los datos). (1 punto)

In [13]:
users = users.withColumnRenamed('flag_molestoso', 'label')
features = users.columns
features.remove('label')
print(features)

['average_stars', 'compliment_cool', 'compliment_cute', 'compliment_funny', 'compliment_hot', 'compliment_list', 'compliment_more', 'compliment_note', 'compliment_photos', 'compliment_plain', 'compliment_profile', 'compliment_writer', 'cool', 'fans', 'funny', 'review_count', 'useful', 'friends_count', 'elite_count']

In [14]:
from pyspark.ml.feature import VectorAssembler

assemble_feats = VectorAssembler(inputCols = features, outputCol = 'assembled_features')
assemble_feats = assemble_feats.transform(users)
assemble_feats = assemble_feats.select('label', 'assembled_features')
assemble_feats.take(1)

[Row(label=1, assembled_features=DenseVector([4.03, 1.0, 0.0, 1.0, 2.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.0, 2.0, 25.0, 5.0, 17.0, 95.0, 84.0, 98.0, 2.0]))]

In [19]:
train, test = assemble_feats.randomSplit([0.7, 0.3])
print("Cantidad de datos de entrenamiento: {}".format(train.count()))
print("Cantidad de datos de validación: {}".format(test.count()))

Cantidad de datos de entrenamiento: 1145795
Cantidad de datos de validación: 491343

#### Entrene tres modelos ( LogisticRegression , GBTClassifier y DecisionTreeClassifier ) sin modificar hiperparámetros que en base a los atributos disponibles en el archivo user.json , clasifique los usuarios molestosos. (6 puntos)

##### Modelo LogisticRegression

In [33]:
from pyspark.ml.classification import LogisticRegression
from pyspark.ml.evaluation import BinaryClassificationEvaluator

logisticRegression_model = LogisticRegression(featuresCol='assembled_features', labelCol='label', predictionCol='molestoso_pred')
logisticRegression_model = logisticRegression_model.fit(train)

evaluator = BinaryClassificationEvaluator()
print("Area Under ROC on Train: {}".format(round(evaluator.evaluate(
    logisticRegression_model.transform(train), {evaluator.metricName: "areaUnderROC"}), 5)))
print("Area Under ROC on Test: {}".format(round(evaluator.evaluate(
    logisticRegression_model.transform(test), {evaluator.metricName: "areaUnderROC"}), 5)))

Area Under ROC on Train: 0.99963
Area Under ROC on Test: 0.99964

##### Modelo GBTClassifier

In [34]:
from pyspark.ml.classification import GBTClassifier

gbt_model = GBTClassifier(featuresCol='assembled_features', labelCol='label', predictionCol='molestoso_pred')
gbt_model = gbt_model.fit(train)

print("Area Under ROC on Train: {}".format(round(evaluator.evaluate(
    gbt_model.transform(train), {evaluator.metricName: "areaUnderROC"}), 5)))
print("Area Under ROC on Test: {}".format(round(evaluator.evaluate(
    gbt_model.transform(test), {evaluator.metricName: "areaUnderROC"}), 5)))

Area Under ROC on Train: 0.99994
Area Under ROC on Test: 0.99992

##### Modelo DecisionTreeClassifier

In [35]:
from pyspark.ml.classification import DecisionTreeClassifier

decisionTree_model = DecisionTreeClassifier(featuresCol='assembled_features', labelCol='label', predictionCol='molestoso_pred')
decisionTree_model = decisionTree_model.fit(train)

print("Area Under ROC on Train: {}".format(round(evaluator.evaluate(
    decisionTree_model.transform(train), {evaluator.metricName: "areaUnderROC"}), 5)))
print("Area Under ROC on Test: {}".format(round(evaluator.evaluate(
    decisionTree_model.transform(test), {evaluator.metricName: "areaUnderROC"}), 5)))

Area Under ROC on Train: 0.9975
Area Under ROC on Test: 0.9975

#### Reporte cuál es el mejor modelo en base a la métrica AUC. (1 punto)
* Considerando la métrica AUC tanto en los conjuntos de datos de entrenamiento y validación, el mejor modelo es el GBTClassifier.

#### Identifique cuales son los principales atributos asociados a un usuario molestoso y repórtelos. (2 puntos)

In [36]:
gbt_model.featureImportances

SparseVector(19, {0: 0.0019, 1: 0.0001, 2: 0.0, 4: 0.0014, 5: 0.0, 6: 0.0002, 7: 0.0, 8: 0.0001, 9: 0.0, 10: 0.0, 11: 0.0003, 12: 0.0028, 13: 0.0446, 14: 0.0005, 15: 0.8955, 16: 0.0432, 17: 0.0008, 18: 0.0087})

Los principales atributos asociados a un usuario molestoso son:
    * review_count con un 0.8955
    * fans con un 0.0446
    * useful con un 0.0432

### Ejercicio 2: Identificando la probabilidad de cierre de un servicio (14 puntos)

#### Implemente el esquema de recodificación. (2 puntos)

In [67]:
from pyspark.sql.functions import col

# 1 if business accepts insurance, 0 otherwise
business = business\
            .withColumn('accepts_insurance',
                        when((col('attributes.AcceptsInsurance') == 'True')\
                             | (col('attributes.AcceptsInsurance') == "\'True\'")\
                             | (col('attributes.AcceptsInsurance') == "u\'True\'"), 1)\
                        .otherwise(0))

# 1 if business is age-restricted, 0 otherwise
business = business\
            .withColumn('all_ages_allowed',
                        when((col('attributes.AgesAllowed') == 'allages')\
                             | (col('attributes.AgesAllowed') == "\'allages\'")\
                             | (col('attributes.AgesAllowed') == "u\'allages\'"), 1)\
                        .otherwise(0))

# 1 if business allows alcohol consumption, 0 otherwise
business = business\
            .withColumn('alcohol_consumption',
                        when((col('attributes.Alcohol') == 'beer_and_wine')\
                             | (col('attributes.Alcohol') == "\'beer_and_wine\'")\
                             | (col('attributes.Alcohol') == "u\'beer_and_wine\'")\
                             | (col('attributes.Alcohol') == 'full_bar')\
                             | (col('attributes.Alcohol') == "\'full_bar\'")\
                             | (col('attributes.Alcohol') == "u\'full_bar\'"), 1)\
                        .otherwise(0))

# 1 if business accepts bitcoin as payment method, 0 otherwise
business = business\
            .withColumn('bitcoin_friendly',
                        when((col('attributes.BusinessAcceptsBitcoin') == 'True')\
                             | (col('attributes.BusinessAcceptsBitcoin') == True)\
                             | (col('attributes.BusinessAcceptsBitcoin') == "\'True\'")\
                             | (col('attributes.BusinessAcceptsBitcoin') == "u\'True\'"), 1)\
                        .otherwise(0))

# 1 if business is food related, 0 otherwise
business = business\
            .withColumn('food_related',
                        when((col('categories').rlike('Food'))\
                             | (col('categories').rlike('Restaurants'))\
                             | (col('categories').rlike('Bars')), 1)\
                        .otherwise(0))

# 1 if business is finance related, 0 otherwise
business = business\
            .withColumn('finance_related',
                        when((col('categories').rlike('Banks'))\
                             | (col('categories').rlike('Insurance'))\
                             | (col('categories').rlike('Finance')), 1)\
                        .otherwise(0))

# 1 if business is health related, 0 otherwise
business = business\
            .withColumn('health_related',
                        when((col('categories').rlike('Fitness'))\
                             | (col('categories').rlike('Hospitals'))\
                             | (col('categories').rlike('Health')), 1)\
                        .otherwise(0))

# 1 if business allows smoking, 0 otherwise
business = business\
            .withColumn('smoking',
                        when((col('attributes.Smoking') == '\'yes\'')\
                             | (col('attributes.Smoking') == 'u\'yes\'')\
                             | (col('attributes.Smoking') == 'yes')\
                             | (col('attributes.Smoking') == '\'outdoor\'')\
                             | (col('attributes.Smoking') == 'u\'outdoor\'')\
                             | (col('attributes.Smoking') == 'outdoor'), 1)\
                        .otherwise(0))

# 1 if business offers free wifi to its customers, 0 otherwise
business = business\
            .withColumn('free_wifi',
                        when((col('attributes.WiFi') == '\'free\'')\
                             | (col('attributes.WiFi') == 'u\'free\'')\
                             | (col('attributes.WiFi') == 'free'), 1)\
                        .otherwise(0))

# 1 if business can be considered expensive, 0 otherwise (applies only to restaurants)
business = business\
            .withColumn('expensive_restaurant',
                        when((col('attributes.RestaurantsPriceRange2') == 3)\
                             | (col('attributes.RestaurantsPriceRange2') == 4), 1)\
                        .otherwise(0))

# 1 if business is kid friendly, 0 otherwise
business = business\
            .withColumn('kid_friendly',
                        when((col('attributes.GoodForKids') == 'True')\
                             | (col('attributes.GoodForKids') == True)\
                             | (col('attributes.GoodForKids') == "\'True\'")\
                             | (col('attributes.GoodForKids') == "u\'True\'"), 1)\
                        .otherwise(0))

# 1 if business has tv, 0 otherwise
business = business\
            .withColumn('has_tv',
                        when((col('attributes.HasTV') == 'True')\
                             | (col('attributes.HasTV') == True)\
                             | (col('attributes.HasTV') == "\'True\'")\
                             | (col('attributes.HasTV') == "u\'True\'"), 1)\
                        .otherwise(0))

# 1 if business is dog friendly, 0 otherwise
business = business\
            .withColumn('dog_friendly',
                        when((col('attributes.DogsAllowed') == 'True')\
                             | (col('attributes.DogsAllowed') == True)\
                             | (col('attributes.DogsAllowed') == "\True'\'")\
                             | (col('attributes.DogsAllowed') == "u\'True\'"), 1)\
                        .otherwise(0))

# 1 if business can be considered as loud or very loud, 0 otherwise
business = business\
            .withColumn('loud_place',
                        when((col('attributes.NoiseLevel') == 'loud')\
                             | (col('attributes.NoiseLevel') == "\'loud\'")\
                             | (col('attributes.NoiseLevel') == "u\'loud\'")\
                             | (col('attributes.NoiseLevel') == "very_loud")\
                             | (col('attributes.NoiseLevel') == "\'very_loud\'")\
                             | (col('attributes.NoiseLevel') == "u\'very_loud\'"), 1)\
                        .otherwise(0))

# 1 if business offers happy hours, 0 otherwise (applies only to restaurants)
business = business\
            .withColumn('happy_hour',
                        when((col('attributes.HappyHour') == 'True')\
                             | (col('attributes.HappyHour') == True)\
                             | (col('attributes.HappyHour') == "\'True\'")\
                             | (col('attributes.HappyHour') == "u\'True\'"), 1)\
                        .otherwise(0))

#### Genere la recodificación del vector objetivo. (2 puntos)

In [68]:
# 1 for closed, 0 otherwise
business = business\
            .withColumn('label',
                        when(col('is_open') == 0, 1)\
                        .otherwise(0))

business = business\
            .select('label', 'review_count', 'stars', 'accepts_insurance',
                    'all_ages_allowed', 'alcohol_consumption', 'bitcoin_friendly', 'food_related',
                    'finance_related', 'health_related', 'smoking', 'free_wifi', 'has_tv',
                    'dog_friendly', 'kid_friendly', 'expensive_restaurant', 'loud_place', 'happy_hour')

In [71]:
business.take(1)

[Row(label=1, review_count=5, stars=3.0, accepts_insurance=0, all_ages_allowed=0, alcohol_consumption=0, bitcoin_friendly=0, food_related=0, finance_related=0, health_related=0, smoking=0, free_wifi=0, has_tv=0, dog_friendly=0, kid_friendly=0, expensive_restaurant=0, loud_place=0, happy_hour=0)]

In [70]:
business.printSchema()

root
 |-- label: integer (nullable = false)
 |-- review_count: long (nullable = true)
 |-- stars: double (nullable = true)
 |-- accepts_insurance: integer (nullable = false)
 |-- all_ages_allowed: integer (nullable = false)
 |-- alcohol_consumption: integer (nullable = false)
 |-- bitcoin_friendly: integer (nullable = false)
 |-- food_related: integer (nullable = false)
 |-- finance_related: integer (nullable = false)
 |-- health_related: integer (nullable = false)
 |-- smoking: integer (nullable = false)
 |-- free_wifi: integer (nullable = false)
 |-- has_tv: integer (nullable = false)
 |-- dog_friendly: integer (nullable = false)
 |-- kid_friendly: integer (nullable = false)
 |-- expensive_restaurant: integer (nullable = false)
 |-- loud_place: integer (nullable = false)
 |-- happy_hour: integer (nullable = false)

#### Divida la muestra en conjuntos de entrenamiento (preservando un 70% de los datos) y validación (preservando un 30% de los datos). (1 punto)

In [72]:
features = business.columns
features.remove('label')
print(features)

['review_count', 'stars', 'accepts_insurance', 'all_ages_allowed', 'alcohol_consumption', 'bitcoin_friendly', 'food_related', 'finance_related', 'health_related', 'smoking', 'free_wifi', 'has_tv', 'dog_friendly', 'kid_friendly', 'expensive_restaurant', 'loud_place', 'happy_hour']

In [75]:
from pyspark.ml.feature import VectorAssembler

assemble_feats = VectorAssembler(inputCols = features, outputCol = 'assembled_features')
assemble_feats = assemble_feats.transform(business)
assemble_feats = assemble_feats.select('label', 'assembled_features')
assemble_feats.take(1)

[Row(label=1, assembled_features=SparseVector(17, {0: 5.0, 1: 3.0}))]

In [76]:
train, test = assemble_feats.randomSplit([0.7, 0.3])
print("Cantidad de datos de entrenamiento: {}".format(train.count()))
print("Cantidad de datos de validación: {}".format(test.count()))

Cantidad de datos de entrenamiento: 135015
Cantidad de datos de validación: 57594

#### Entrene tres modelos ( LogisticRegression , GBTClassifier y DecisionTreeClassifier ) sin modificar hiperparámetros que en base a los atributos recodificados del archivo business.json , clasifique aquellos servicios cerrados. (6 puntos)

##### Modelo LogisticRegression

In [77]:
from pyspark.ml.classification import LogisticRegression
from pyspark.ml.evaluation import BinaryClassificationEvaluator

logisticRegression_model = LogisticRegression(featuresCol='assembled_features', labelCol='label', predictionCol='is_closed_pred')
logisticRegression_model = logisticRegression_model.fit(train)

evaluator = BinaryClassificationEvaluator()
print("Area Under ROC on Train: {}".format(round(evaluator.evaluate(
    logisticRegression_model.transform(train), {evaluator.metricName: "areaUnderROC"}), 5)))
print("Area Under ROC on Test: {}".format(round(evaluator.evaluate(
    logisticRegression_model.transform(test), {evaluator.metricName: "areaUnderROC"}), 5)))

Area Under ROC on Train: 0.6941
Area Under ROC on Test: 0.68866

##### Modelo GBTClassifier

In [78]:
from pyspark.ml.classification import GBTClassifier

gbt_model = GBTClassifier(featuresCol='assembled_features', labelCol='label', predictionCol='is_closed_pred')
gbt_model = gbt_model.fit(train)

print("Area Under ROC on Train: {}".format(round(evaluator.evaluate(
    gbt_model.transform(train), {evaluator.metricName: "areaUnderROC"}), 5)))
print("Area Under ROC on Test: {}".format(round(evaluator.evaluate(
    gbt_model.transform(test), {evaluator.metricName: "areaUnderROC"}), 5)))

Area Under ROC on Train: 0.71102
Area Under ROC on Test: 0.70243

##### Modelo DecisionTreeClassifier

In [79]:
from pyspark.ml.classification import DecisionTreeClassifier

decisionTree_model = DecisionTreeClassifier(featuresCol='assembled_features', labelCol='label', predictionCol='is_closed_pred')
decisionTree_model = decisionTree_model.fit(train)

print("Area Under ROC on Train: {}".format(round(evaluator.evaluate(
    decisionTree_model.transform(train), {evaluator.metricName: "areaUnderROC"}), 5)))
print("Area Under ROC on Test: {}".format(round(evaluator.evaluate(
    decisionTree_model.transform(test), {evaluator.metricName: "areaUnderROC"}), 5)))

Area Under ROC on Train: 0.35991
Area Under ROC on Test: 0.36734

#### Reporte cuál es el mejor modelo en base a la métrica AUC. (1 punto)
* Considerando la métrica AUC tanto en los conjuntos de datos de entrenamiento y validación, el mejor modelo es el GBTClassifier.

#### Identifique cuales son los principales atributos asociados al cierre de un servicio. (2 puntos)

In [95]:
gbt_model.featureImportances

SparseVector(17, {0: 0.2164, 1: 0.1688, 2: 0.0658, 3: 0.0009, 4: 0.084, 5: 0.0004, 6: 0.1691, 7: 0.0178, 8: 0.02, 9: 0.0094, 10: 0.0297, 11: 0.0403, 12: 0.012, 13: 0.0782, 14: 0.032, 15: 0.0238, 16: 0.0314})

Los principales atributos asociados a un cierre de servicio son:

* review_count con un 0.2164
* food_related con un 0.1691
* stars con un 0.1688